extension/src/tests/contract.rs

```diff
@@ -0,0 +1,164 @@
+			let contract = instantiate(CONTRACT.clone());
+			let dispatch_result = call(
+				contract,
+				DispatchContractNoopFuncId::get(),
+				RuntimeCall::System(Call::remark_with_event { remark: "pop".as_bytes().to_vec() }),
+				GAS_LIMIT,
+			);
+			assert_eq!(
+				dispatch_result.result,
+				Err(Module(ModuleError {
+					index: 0,
+					error: [5, 0, 0, 0],
+					message: Some("CallFiltered")
+				}))
+			);
+		});
+	}
+
+	#[test]
+	fn dispatch_call_return_error_fails() {
+		new_test_ext().execute_with(|| {
+			// Instantiate a new contract.
+			let contract = instantiate(CONTRACT.clone());
+			let dispatch_result = call(
+				contract,
+				DispatchContractFuncId::get(),
+				// `set_code` requires root origin, expect throwing error.
+				RuntimeCall::System(Call::set_code { code: "pop".as_bytes().to_vec() }),
+				GAS_LIMIT,
+			);
+			assert_eq!(dispatch_result.result.err(), Some(BadOrigin))
+		})
+	}
+}
+
+#[cfg(test)]
+mod read_state_tests {
+	use super::*;
+
+	#[test]
+	fn read_state_works() {
+		new_test_ext().execute_with(|| {
+			// Instantiate a new contract.
+			let contract = instantiate(CONTRACT.clone());
+			// Successfully return data.
+			let read_result =
+				call(contract, ReadContractFuncId::get(), RuntimeRead::Ping, GAS_LIMIT);
+			let return_value = read_result.result.unwrap();
+			let decoded = <Result<Vec<u8>, u32>>::decode(&mut &return_value.data[1..]).unwrap();
+			let result = Ok("pop".as_bytes().to_vec());
+			assert_eq!(decoded, result);
+		});
+	}
+
+	#[test]
+	fn read_state_filtering_noop_fails() {
+		new_test_ext().execute_with(|| {
+			// Instantiate a new contract.
+			let contract = instantiate(CONTRACT.clone());
+			// Successfully return data.
+			let read_result =
+				call(contract, ReadContractNoopFuncId::get(), RuntimeRead::Ping, GAS_LIMIT);
+			assert_eq!(
+				read_result.result,
+				Err(Module(ModuleError {
+					index: 0,
+					error: [5, 0, 0, 0],
+					message: Some("CallFiltered")
+				}))
+			);
+		});
+	}
+
+	#[test]
+	fn read_state_invalid_input_fails() {
+		new_test_ext().execute_with(|| {
+			// Instantiate a new contract.
+			let contract = instantiate(CONTRACT.clone());
+			let read_result = call(contract, ReadExtFuncId::get(), 99u8, GAS_LIMIT);
+			let expected: DispatchError = pallet_contracts::Error::<Test>::DecodingFailed.into();
+			// Make sure the error is passed through the error converter.
+			let error =
+				<() as ErrorConverter>::convert(expected, &mock::MockEnvironment::default()).err();
+			assert_eq!(read_result.result.err(), error);
+		})
+	}
+}
+
+#[test]
+fn noop_function_works() {
+	new_test_ext().execute_with(|| {
+		// Instantiate a new contract.
+		let contract = instantiate(CONTRACT.clone());
+		let noop_result = call(contract, NoopFuncId::get(), (), GAS_LIMIT);
+		// Successfully return data.
+		let return_value = noop_result.result.unwrap();
+		let decoded = <Result<Vec<u8>, u32>>::decode(&mut &return_value.data[..]).unwrap();
```

Comment:
why does this include in the first byte, while a few tests above it excludes it?

Reply:
Confirmed. The Noop function called does nothing and just returns `Ok(RetVal::Converging(0))` so there is no data written.